### PR TITLE
gh-77273: Better bytecodes for f-strings

### PR DESCRIPTION
**Pull request by markshannon**
Simplify and speed up interpreter for f-strings.

tldr; Do work in the compiler instead of the interpreter.

Splits the FORMAT_VALUE opcode into CONVERT_VALUE, FORMAT_SIMPLE and FORMAT_WITH_SPEC.
The compiler can then emit the optimal sequence for each format expression.

The three new opcodes are much simpler than the one they replace, actually removing a few lines of code from the interpreter.

The CONVERT_VALUE is emitted only if a conversion is present.
In the standard library only (approx) 20% of f-strings include a conversion.

FORMAT_WITH_SPEC is emitted if a format specifier is present.
Otherwise FORMAT_SIMPLE  is emitted. 
In the standard library about 70% of format expressions can be formatted with just the simple (and fast)
FORMAT_SIMPLE opcode.

https://bugs.python.org/issue33092

* Issue: gh-77273

### CODE REVIEW

**Comment by alimcmaster1** [comment]
Any update on this @taleinat suggested sharing benchmarks here https://bugs.python.org/issue33092. 

Otherwise due to lack of activity I think we can close this.

**Comment by furkanonder** [comment]
CC: @rhettinger @ambv

**Comment by markshannon** [comment]
Updated to use the new bytecode definition format.

Removes the conditional stack effect, simplifying tooling around these instructions.

**Comment by iritkatriel** [comment]
Looks like you missed out the doc update.

File: Lib/dis.py [hunk 1]
```diff
@@ -508,13 +503,9 @@ def _get_instructions_bytes(code, varname_from_oparg=None,
...
-            elif deop == FORMAT_VALUE:
-                argval, argrepr = FORMAT_VALUE_CONVERTERS[arg & 0x3]
-                argval = (argval, bool(arg & 0x4))
-                if argval[1]:
-                    if argrepr:
-                        argrepr += ', '
-                    argrepr += 'with format'
+            elif deop == CONVERT_VALUE:
+                argval = (None, str, repr, ascii)[arg]
+                argrepr = ('', 'str', 'repr', 'ascii')[arg]
```
Comment (iritkatriel) [inline Lib/dis.py:579]: since oparg 0 isn't a thing (right?), maybe write this as:

```suggestion
                argval = (_, str, repr, ascii)[arg]
                argrepr = (_, 'str', 'repr', 'ascii')[arg]
```

**Comment by markshannon** [comment]
https://github.com/python/cpython/issues/105678

File: Lib/dis.py (hunk 1 above)
Comment (markshannon) [reply Lib/dis.py:579]: What is `_`? We might not use the 0th item, but it still needs to be defined.

Comment (iritkatriel) [reply Lib/dis.py:579]: IIUC _ is the convention for ‘unused value’.

Comment (markshannon) [reply Lib/dis.py:579]: For lvalues, sure, but this is an rvalue.

**Review by iritkatriel — APPROVED** [review]
